File: day13/src/main/python/solution.py
def read_input():
    f = open('../resources/input', 'r')
    coordinates = set()
    instructions = []
    complete = False
    for line in f.readlines():
        if line.strip() == '':
            complete = True
        elif complete:
            d, i = line.strip().split("=")
            instructions.append((d[-1], int(i)))
        else:
            x, y = line.strip().split(',')
            coordinates.add((int(x), int(y)))
    return coordinates, instructions


def do_fold(coordinates, instruction):
    return set(map(lambda c: coordinate_fold(c, instruction), coordinates))


def coordinate_fold(coordinate, instruction):
    d, i = instruction
    x, y = coordinate
    if d == 'x' and x > i:
        return 2 * i - x, y
    elif d == 'y' and y > i:
        return x, 2 * i - y
    return x, y


def fold_result(coordinates, instructions):
    current = coordinates.copy()
    for instruction in instructions:
        current = do_fold(current, instruction)
    result = ""
    for y in range(0, 6):
        for x in range(0, 40):
            result += '#' if (x, y) in current else '.'
        result += '\n'
    return result


if __name__ == "__main__":
    initial_coordinates, instructin_list = read_input()
    print("part 1")
    print(len(do_fold(initial_coordinates, instructin_list[0])))
    print("part 2")
    print(fold_result(initial_coordinates, instructin_list))

File: day13/src/main/cpp/solution.cpp
#include <fstream>
#include <iostream>
#include <set>
#include <string>
#include <vector>

using namespace std;
#define coordinate pair<int, int>
#define instruction pair<string, int>

pair<set<coordinate>, vector<instruction>> readInputFile() {
  ifstream file("../resources/input");
  set<coordinate> coordinates;
  vector<instruction> instructions;
  bool instr = false;
  string s, d;
  int i, delim, x, y;
  while (getline(file, s)) {
    if (s == "") {
      instr = true;
    } else if (instr) {
      d = s.substr(11, 1);
      i = stoi(s.substr(13, 9));
      instructions.push_back(instruction(d, i));
    } else {
      delim = s.find(",");
      x = stoi(s.substr(0, delim));
      y = stoi(s.substr(delim + 1, s.length()));
      coordinates.insert(coordinate(x, y));
    }
  }
  return pair(coordinates, instructions);
}

coordinate foldCoordinate(coordinate c, instruction i) {
  int x = c.first;
  int y = c.second;
  if (i.first == "x" && x > i.second)
    return coordinate(2 * i.second - x, y);
  if (i.first == "y" && y > i.second)
    return coordinate(x, 2 * i.second - y);
  return c;
}

set<coordinate> foldPage(set<coordinate> &coordinates, instruction i) {
  set<coordinate> result;
  for (auto const &c : coordinates) {
    result.insert(foldCoordinate(c, i));
  }
  return result;
}

void printFoldedPages(set<coordinate> coordinates,
                      vector<instruction> instructions) {
  set<coordinate> current = coordinates;
  for (auto const &i : instructions) {
    current = foldPage(current, i);
  }
  for (int y = 0; y < 6; y++) {
    for (int x = 0; x < 40; x++) {
      if (current.find(coordinate(x, y)) == current.end()) {
        cout << ".";
      } else {
        cout << "#";
      }
    }
    cout << endl;
  }
}

int main() {
  auto input = readInputFile();
  cout << "part 1" << endl
       << foldPage(input.first, input.second[0]).size() << endl
       << "part 2" << endl;
  printFoldedPages(input.first, input.second);
  return 0;
}
